storage-node/src/services/sync/tasks.ts

```diff
@@ -77,39 +72,62 @@ export class DeleteLocalFileTask implements SyncTask {
  * Download the file from the remote storage node to the local storage.
  */
 export class DownloadFileTask implements SyncTask {
-  dataObjectId: string
-  expectedHash?: string
-  uploadsDirectory: string
-  tempDirectory: string
-  url: string
-  downloadTimeout: number
-  hostId: string
+  operatorUrls: string[]
 
   constructor(
-    baseUrl: string,
-    dataObjectId: string,
-    expectedHash: string | undefined,
-    uploadsDirectory: string,
-    tempDirectory: string,
-    downloadTimeout: number,
-    hostId: string
+    baseUrls: string[],
+    private dataObjectId: string,
+    private expectedHash: string,
+    private uploadsDirectory: string,
+    private tempDirectory: string,
+    private downloadTimeout: number,
+    private hostId: string
   ) {
-    this.dataObjectId = dataObjectId
-    this.expectedHash = expectedHash
-    this.uploadsDirectory = uploadsDirectory
-    this.tempDirectory = tempDirectory
-    this.downloadTimeout = downloadTimeout
-    this.url = urljoin(baseUrl, 'api/v1/files', dataObjectId)
-    this.hostId = hostId
+    this.operatorUrls = baseUrls.map((baseUrl) => urljoin(baseUrl, 'api/v1/files', dataObjectId))
   }
 
   description(): string {
-    return `Sync - downloading file: ${this.url} to ${this.uploadsDirectory} ....`
+    return `Sync - Trying for download of: ${this.dataObjectId} ....`
   }
 
   async execute(): Promise<void> {
+    // Create an array of operator URL indices to maintain a random URL choice
+    // cannot use the original array because we shouldn't modify the original data.
+    // And cloning it seems like a heavy operation.
+    const operatorUrlIndices: number[] = [...Array(this.operatorUrls.length).keys()]
```

Comment:
I can suggest a slightly cleaner approach. Start with an array already containing random order of the indices using a combination of _.range() and _.shuffle(). You could then just shift or pop the array at the start of the loop. Or the while loop then can be a simply a `for..each` iteration over the array, without needing to modify it.

`const operatorUrlIndices: number[] = _.shuffle(_.range(this.operatorUrls.length))`